### PR TITLE
Support MPIJob managedBy feature for the MultiKueue

### LINKED ISSUE #3257: Support MPIJob managedBy feature for the MultiKueue
**What would you like to be added**:
If we want to use the current MPIJob MultiKueue feature, we need to uninstall the mpi-operator from the management cluster.
But, since the [mpi-operator v0.6.0](https://github.com/kubeflow/mpi-operator/releases/tag/v0.6.0), we started to support the managedBy feature similar to the batch/v1 Job.

Hence, we want to support it by implementing in the following:

1. Implement the `IsJobManagedByKueue`: https://github.com/kubernetes-sigs/kueue/blob/4199c9dd9ce89636eb4e72f5cebb3c9adfba3f0c/pkg/controller/jobs/mpijob/mpijob_multikueue_adapter.go#L89-L91
2. Implement the defaulting webhoooks similar to [Job](https://github.com/kubernetes-sigs/kueue/blob/4199c9dd9ce89636eb4e72f5cebb3c9adfba3f0c/pkg/controller/jobs/job/job_webhook.go#L84-L102) and [JobSet](https://github.com/kubernetes-sigs/kueue/blob/4199c9dd9ce89636eb4e72f5cebb3c9adfba3f0c/pkg/controller/jobs/jobset/jobset_webhook.go#L72-L91): 

And other needed implementations to support it if it exists.

**Why is this needed**:
The MPIJob with managed feature allows us to install the mpi-operator to management and worker clusters so that we can easily introduce multi-cluster MPIJob dispatching and give the possibility of running the MPIJob in the management cluster as well.

**Completion requirements**:

This enhancement requires the following artifacts:

- [ ] Design doc
- [ ] API change
- [x] Docs update

The artifacts should be linked in subsequent comments.

### PR DESCRIPTION
**Pull request by mszadkow**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Support MPIJob managedBy feature for the MultiKueue.
Allow for installing `mpi-operator` in management cluster instead of only crds. 

#### Which issue(s) this PR fixes:

Fixes #3257 

#### Does this PR introduce a user-facing change?

```release-note
MultiKueue: Add support for  MPIJob  `spec.runPolicy.managedBy` field
```

### CODE REVIEW

File: test/integration/multikueue/multikueue_test.go [hunk 1]
```diff
@@ -1050,8 +1050,62 @@ var _ = ginkgo.Describe("Multikueue", ginkgo.Ordered, ginkgo.ContinueOnFailure,
...
 
 	ginkgo.It("Should run a MPIJob on worker if admitted", func() {
+		admission := utiltesting.MakeAdmission(managerCq.Name).PodSets(
+			kueue.PodSetAssignment{
+				Name: "launcher",
+			}, kueue.PodSetAssignment{
+				Name: "worker",
+			},
+		)
+		mpijobNoManagedBy := testingmpijob.MakeMPIJob("mpijob2", managerNs.Name).
```
Comment (mbobrovskyi) [inline test/integration/multikueue/multikueue_test.go:1060]: Can we move it to separate `It`?

**Review by mbobrovskyi — COMMENTED** [review]
Should we also implement defaulting webhoooks?

File: test/integration/multikueue/multikueue_test.go (hunk 1 above)
Comment (mszadkow) [reply test/integration/multikueue/multikueue_test.go:1060]: we could I guess, I wanted to reduce testing time, keeping it here

**Comment by mszadkow** [comment]
> Should we also implement defaulting webhoooks?

Good question, I guess it's true...

**Comment by mimowo** [comment]
Yes, we want the webhook defaulting too.

**Comment by tenzen-y** [comment]
As I mentioned in the issue, let's implement the dedicated webhook instead of base: https://github.com/kubernetes-sigs/kueue/blob/4199c9dd9ce89636eb4e72f5cebb3c9adfba3f0c/pkg/controller/jobs/mpijob/mpijob_controller.go#L44

**Comment by mszadkow** [comment]
e2e test fails due to: https://github.com/kubernetes-sigs/kueue/issues/3299
Except for that ready for review
@mbobrovskyi @trasc @vladikkuzn

File: test/integration/multikueue/multikueue_test.go (hunk 1 above)
Comment (mszadkow) [reply test/integration/multikueue/multikueue_test.go:1060]: done

**Review by tenzen-y — COMMENTED** [review]
@mszadkow Thank you for driving this feature across kubeflow and Kueue!
I left a few nits comments.

File: pkg/controller/jobs/mpijob/mpijob_multikueue_adapter.go [hunk 2]
```diff
@@ -86,7 +90,16 @@ func (b *multikueueAdapter) KeepAdmissionCheckPending() bool {
 	return false
 }
 
-func (b *multikueueAdapter) IsJobManagedByKueue(context.Context, client.Client, types.NamespacedName) (bool, string, error) {
+func (b *multikueueAdapter) IsJobManagedByKueue(ctx context.Context, remoteClient client.Client, key types.NamespacedName) (bool, string, error) {
```
Comment (trasc) [inline pkg/controller/jobs/mpijob/mpijob_multikueue_adapter.go:93]: ` remoteClient` .. isn't it usually the local client?

**Review by trasc — COMMENTED** [review]
one nit

File: pkg/controller/jobs/mpijob/mpijob_webhook.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+		}
+	}
+
+	return nil
+}
+
+func canDefaultManagedBy(mpiJobSpecManagedBy *string) bool {
+	return features.Enabled(features.MultiKueue) &&
+		(mpiJobSpecManagedBy == nil || *mpiJobSpecManagedBy == v2beta1.KubeflowJobController)
+}
```
[the remark this answers is not in the document]
Comment (mszadkow) [reply pkg/controller/jobs/mpijob/mpijob_webhook.go:100]: A follow up, fair deal ;)

File: pkg/controller/jobs/mpijob/mpijob_multikueue_adapter_test.go [hunk 4]
```diff
@@ -116,16 +116,53 @@ func TestMultikueueAdapter(t *testing.T) {
 				return adapter.DeleteRemoteObject(ctx, workerClient, types.NamespacedName{Name: "mpijob1", Namespace: TestNamespace})
 			},
 		},
```
[the remark this answers is not in the document]
Comment (mszadkow) [reply pkg/controller/jobs/mpijob/mpijob_multikueue_adapter_test.go:118]: Added

**Review by tenzen-y — COMMENTED** [review]
Thanks!
@mszadkow Could you open PR to update the documentation?
/lgtm
/approve

**Comment by mszadkow** [comment]
> Thanks! @mszadkow Could you open PR to update the documentation? /lgtm /approve

Thank you!
Yes, I will open it.

**Comment by mszadkow** [comment]
https://github.com/kubernetes-sigs/kueue/issues/3313

**Comment by mszadkow** [comment]
https://github.com/kubernetes-sigs/kueue/pull/3316 - docs PR

**Pull request merged**